dali/imgcodec/tools/imagemagick_test.cc

```diff
@@ -0,0 +1,283 @@
+  stream << "FAIL" << "\t" << filename << ": " << message << std::endl;
+}
+
+void ok(std::ostream &stream, const std::string &filename) {
+  stream << "OK" << "\t" << filename << std::endl;
+}
+
+class ImgcodecTester {
+ public:
+  ImgcodecTester() {
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("jpeg", std::make_shared<JpegParser>()));
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("png", std::make_shared<PngParser>()));
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("bmp", std::make_shared<BmpParser>()));
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("tiff", std::make_shared<TiffParser>()));
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("pnm", std::make_shared<PnmParser>()));
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("jpeg2000", std::make_shared<Jpeg2000Parser>()));
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("webp", std::make_shared<WebpParser>()));
+  }
+
+  std::optional<TensorShape<>> shape_of(const std::string &filename) const {
+    auto img = ImageSource::FromFilename(filename);
+    auto fmt = this->format_registry_.GetImageFormat(&img);
+    if (fmt == nullptr) return {};
+    auto image_info = fmt->Parser()->GetInfo(&img);
+    return image_info.shape;
+  }
+
+  ImageFormatRegistry format_registry_;
+};
+
+struct Env {
+  std::string identify_path;
+  std::string directory;
+  std::regex filter;
+  bool print;
+  unsigned batch;
+  unsigned jobs;
+  ImgcodecTester imgcodec_tester = {};
+  std::mutex mutex = {};
+};
+
+std::vector<std::string> get_batch(const Env &env,
+                                   std::filesystem::recursive_directory_iterator &it) {
+  std::vector<std::string> filenames;
+  filenames.reserve(env.batch);
+  while (it != std::filesystem::end(it) && filenames.size() < env.batch) {
+    const auto& entry = *(it++);
+    if (entry.is_regular_file()) {
+      const auto path = entry.path().string();
+      if (std::regex_match(path, env.filter))
+        filenames.push_back(path);
+    }
+  }
+  return filenames;
+}
+
+std::optional<TensorShape<>> scan_imagemagick_shape(FILE *pipe) {
+  int w, h, c;
+  char tmp[16];
+  std::string colors;
+  if (fscanf(pipe, "%d %d %16s", &w, &h, tmp) != 3) return {};
+  colors = tmp;
+
+  if (colors == "srgb" || colors == "rgb") {
+    c = 3;
+  } else if (colors == "srgba" || colors == "rgba" || colors == "cmyk") {
+    c = 4;
+  } else if (colors == "gray") {
+    c = 1;
+  } else {
+    return {};
+  }
+
+  TensorShape<> imagemagick_shape = {h, w, c};
+  return imagemagick_shape;
+}
+
+void process(Env &env, std::vector<std::string> filenames) {
+  std::ostringstream log, out;
+
+  std::ostringstream cmd;
+  cmd << env.identify_path << " -quiet -format  \"%w %h %[channels]\\n\"";
```

Comment:
Here you're accessing env.identify_path from many threads which is data race.
Below you're taking mutex which protects really nothing.

Reply:
But I only read the identify_path, no writes are possible. Is it still insecure?

And the mutex below made no sense as @banasraf pointed out, so it was removed

Reply:
Sure no conflicting actions here, not a data-race. My bad.